### PR TITLE
Fix decide_worker picking a closing worker

### LINKED ISSUE #6311: CI failing with `test_submit_after_failed_worker_async`
Seeing a [CI failure]( https://github.com/dask/distributed/runs/6360987363?check_suite_focus=true ) with `test_submit_after_failed_worker_async`.

### LINKED ISSUE #8019: AssertionError in decide_worker_non_rootish
I got a random CI failure in test_submit_after_failed_worker_async:
https://github.com/dask/distributed/actions/runs/5610919845/jobs/10266547694?pr=8013

I understand this is separate from https://github.com/dask/distributed/issues/6311 as the error is different:

```
2023-07-20 12:29:03,679 - distributed.scheduler - ERROR - Error transitioning 'sum-1662e8752439e04be62efae3b3703604' from 'waiting' to 'processing'
Traceback (most recent call last):
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 1918, in _transition
    recommendations, client_msgs, worker_msgs = func(
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 2292, in transition_waiting_processing
    if not (ws := self.decide_worker_non_rootish(ts)):
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 2268, in decide_worker_non_rootish
    assert ws in self.running, (ws, self.running)
AssertionError: (<WorkerState 'tcp://127.0.0.1:34037', status: closing, memory: 4, processing: 0>, {<WorkerState 'tcp://127.0.0.1:45709', name: 1, status: running, memory: 3, processing: 0>, <WorkerState 'tcp://127.0.0.1:34071', name: 0, status: running, memory: 3, processing: 0>})
2023-07-20 12:29:03,681 - distributed.nanny - INFO - Worker closed
2023-07-20 12:29:03,687 - distributed.scheduler - INFO - Remove client Client-026cc191-26f9-11ee-8805-000d3ae2e24a
2023-07-20 12:29:03,688 - distributed.scheduler - INFO - Remove worker <WorkerState 'tcp://127.0.0.1:34037', status: closing, memory: 0, processing: 0> (stimulus_id='handle-worker-cleanup-1689856143.6881266')
2023-07-20 12:29:03,689 - distributed.scheduler - INFO - Close client connection: Client-026cc191-26f9-11ee-8805-000d3ae2e24a
2023-07-20 12:29:03,689 - distributed.core - ERROR - Exception while handling op register-client
Traceback (most recent call last):
  File "/home/runner/work/distributed/distributed/distributed/core.py", line 930, in _handle_comm
    result = await result
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 5478, in add_client
    await self.handle_stream(comm=comm, extra={"client": client})
  File "/home/runner/work/distributed/distributed/distributed/core.py", line 1014, in handle_stream
    handler(**merge(extra, msg))
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 4488, in update_graph
    self.transitions(recommendations, stimulus_id)
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 7566, in transitions
    self._transitions(recommendations, client_msgs, worker_msgs, stimulus_id)
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 2024, in _transitions
    new_recs, new_cmsgs, new_wmsgs = self._transition(key, finish, stimulus_id)
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 1918, in _transition
    recommendations, client_msgs, worker_msgs = func(
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 2292, in transition_waiting_processing
    if not (ws := self.decide_worker_non_rootish(ts)):
  File "/home/runner/work/distributed/distributed/distributed/scheduler.py", line 2268, in decide_worker_non_rootish
    assert ws in self.running, (ws, self.running)
AssertionError: (<WorkerState 'tcp://127.0.0.1:34037', status: closed, memory: 0, processing: 0>, {<WorkerState 'tcp://127.0.0.1:45709', name: 1, status: running, memory: 0, processing: 0>, <WorkerState 'tcp://127.0.0.1:34071', name: 0, status: running, memory: 0, processing: 0>})
```

### PR DESCRIPTION
**Pull request by crusaderky**
- Closes #6311
- Closes #8019

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -8205,6 +8202,7 @@ def decide_worker(
         candidates = set(all_workers)
     else:
         candidates = {wws for dts in ts.dependencies for wws in dts.who_has}
+        candidates &= all_workers
```
Comment (crusaderky) [inline distributed/scheduler.py:8205]: This fixes #8019

File: distributed/tests/test_failed_workers.py [hunk 2]
```diff
@@ -48,21 +49,57 @@ def test_submit_after_failed_worker_sync(loop):
...
+            in_update_graph.set()
+            await async_poll_for(
+                lambda: b_ws.status == Status.closing, timeout=5, period=0
+            )
+            s.update_graph(*args, **kwargs)
+            nonlocal done_update_graph
+            done_update_graph = True
+
+        s.stream_handlers["update-graph"] = update_graph
+        await in_update_graph.wait()
```
Comment (fjetter) [inline distributed/tests/test_failed_workers.py:94]: Given the problem, I would've expected something like


```python

def block(arg, enter, exit):
    enter.set()
    exit.wait()
    return arg

enter = Event()
exit = Event()
enter2 = Event()
exit2 = Event()

d1 = c.submit(inc, 0, key='d1', workers=["A"])
d2 = c.submit(block, 1, enter=enter, exit=exit, key='d2', workers=["B"])

x = c.submit(sum, [d1, d2], key='x')
block_executor = c.submit(block, None, enter=enter2, exit=exit2, key='x', workers=["B"])

await enter.wait()
await enter2.wait()
await asyncio.gather([
    exit.set(),
    B.close()
])
```

I haven't tested the above and it may still need fine tuning but I would expect something like this to trigger the condition you're talking about. d2 completes while B is closing s.t. when d2 finishes, x is transitioned while B is still closing.

I don't entirely understand why we need update_graph to trigger this condition

Comment (fjetter) [reply distributed/tests/test_failed_workers.py:94]: Especially considering that update_graph (right now) is still synchronous.

Comment (crusaderky) [reply distributed/tests/test_failed_workers.py:94]: I'm not too happy with my test, either. But I don't think your suggestion works (read below).

What we're trying to test:
- `decide_worker_non_rootish(ts)` is called on a task with `workers=[b.address], allow_other_workers=True`,
- with the dependencies of the task partially on a and partially on b,
- while b is in `closing` status, and 
- with `decide_worker(ts, valid_workers=set(), all_workers={a})` that would pick b from `ts.dependencies[...].who_has`, due to less dependency bytes needing to be transferred to it, but instead picks a because b is not in all_workers.

The problem with writing the test is that we need to time `update_graph` to land *exactly* during the 1-2 event loop cycles while the worker is in `closing` status.
The worker transitions from `closing` to being removed when the batched comms collapse, here:
https://github.com/dask/distributed/blob/f0303aaff3a22fd00660312cfa3f74a1fa178938/distributed/scheduler.py#L5700-L5705

Alternatively to monkey-patching update-graph, I could have
- monkey-patched Scheduler.remove_worker. In hindsight that's a better idea; I'll have a look at it now.
- synchronously call update_graph directly on the scheduler and update the state on the client by hand (complicated and brittle).

## In your code:
I think that the scheduler will never receive the task-finished message for d2, since it's a whole 2 cycles of event loop after Worker.close collapses the batched comms.

I'm not sure why you think the scheduler should receive `{op: task-finished, key: d2}` deterministically after `{op: worker-status-change, worker: <b>, status: closing}`, but deterministically before the collapse of the TCP channel?

Also I can't understand what's the purpose of `block_executor`.

File: distributed/tests/test_failed_workers.py [hunk 3]
```diff
@@ -48,21 +49,59 @@ def test_submit_after_failed_worker_sync(loop):
...
+        orig_remove_worker = s.remove_worker
+        in_remove_worker = asyncio.Event()
+        wait_remove_worker = asyncio.Event()
+
+        async def remove_worker(*args, **kwargs):
+            in_remove_worker.set()
+            await wait_remove_worker.wait()
+            return await orig_remove_worker(*args, **kwargs)
+
+        monkeypatch.setattr(s, "remove_worker", remove_worker)
```
Comment (fjetter) [inline distributed/tests/test_failed_workers.py:86]: I'm OK-ish with using monkeypatch here. However, just for the sake of prosperity, there is also a way to use our RPC mechanism more naturally. Essentially you want to intercept the point in time just when a request handler is called. You can make this very explitc

```python
async def new_remove_worker_handler_with_events(self, *args, **kwargs):
    in_remove_worker.set()
    await wait_remove_worker.wait()
    return await self.remove_worker(*args, **kwargs)
s.handlers['unregister'] = new_remove_worker_handler_with_events`
```
Semantically, this overrides the `unregister` handler and replaces it with a new handler.
However, in the end, it's the same thing just the way the patch is installed is different.

File: distributed/tests/test_failed_workers.py [hunk 4]
```diff
@@ -48,21 +49,59 @@ def test_submit_after_failed_worker_sync(loop):
...
+    config={"distributed.comm.timeouts.connect": "1s"},
+)
+async def test_submit_after_failed_worker_async(
+    c, s, a, b, x_on_failed, y_on_failed, when, monkeypatch
+):
+    a_ws = s.workers[a.address]
+
+    L = c.map(
+        inc,
+        range(10),
```
Comment (fjetter) [inline distributed/tests/test_failed_workers.py:67]: Do we actually need a map for this? This feels much more difficult to control than if we used single tasks with specific placement

File: distributed/scheduler.py [hunk 5]
```diff
@@ -2220,9 +2220,6 @@ def decide_worker_non_rootish(self, ts: TaskState) -> WorkerState | None:
 
         valid_workers = self.valid_workers(ts)
         if valid_workers is None and len(self.running) < len(self.workers):
-            if not self.running:
-                return None
```
Comment (fjetter) [inline distributed/scheduler.py:2224]: Is this related? At least the new test doesn't seem to care about this.

File: distributed/scheduler.py (hunk 1 above)
Comment (fjetter) [reply distributed/scheduler.py:8205]: FWIW I think this is a situation where an actual `decide_worker` unit test would be appropriate

File: distributed/tests/test_failed_workers.py [hunk 6]
```diff
@@ -48,21 +49,59 @@ def test_submit_after_failed_worker_sync(loop):
...
+    total = c.submit(
+        sum,
+        L,
+        key="y",
+        workers=[b.address if y_on_failed else a.address],
+        allow_other_workers=True,
+    )
+
+    await wait_for_state("y", "processing", s, interval=0)
+    assert s.tasks["y"].processing_on is a_ws
```
Comment (fjetter) [inline distributed/tests/test_failed_workers.py:100]: I feel this is a bit too specific. We don't _actually_ care about this specific state too much.

```suggestion
    while "y" not in s.tasks:
        await asyncio.sleep(0.05)
```

does the trick as well and properly triggers the assert of https://github.com/dask/distributed/issues/8019

or even better without any sleeping

```python

    def blocked_sum(arg, event):
        event.set()
        return sum(arg)

    event = Event()
    total = c.submit(
        blocked_sum,
        L,
        key="y",
        event=event,
        workers=[b.address if y_on_failed else a.address],
        allow_other_workers=True,
    )

    await event.wait()
```

I'm typically discouraging usage of `wait_for_state`, particularly with very low intervals since this can actually flood the event loop and cause weird issues like the following https://github.com/dask/distributed/issues/6080

File: distributed/tests/test_failed_workers.py (hunk 2 above)
Comment (fjetter) [reply distributed/tests/test_failed_workers.py]: >  but deterministically before the collapse of the TCP channel?

I never said it would do so deterministically. I said I would expect _something like this_ to trigger the condition you described. I also said it would require more fine tuning like distributing nbytes properly and maybe introduce an event somewhere.

File: distributed/tests/test_failed_workers.py (hunk 3 above)
Comment (crusaderky) [reply distributed/tests/test_failed_workers.py:86]: We're arriving here from the `unregister` handler. We're arriving from 
https://github.com/dask/distributed/blob/f0303aaff3a22fd00660312cfa3f74a1fa178938/distributed/scheduler.py#L5700-L5705

File: distributed/tests/test_failed_workers.py (hunk 4 above)
Comment (crusaderky) [reply distributed/tests/test_failed_workers.py:67]: You're right. Simplified.

File: distributed/scheduler.py (hunk 5 above)
Comment (crusaderky) [reply distributed/scheduler.py:2224]: It's unreachable because the same condition is already tested on line 2218

**Comment by crusaderky** [comment]
Code review comments have been addressed

**Review by fjetter — APPROVED** [review]
